src/osp/coordinates.h

```diff
@@ -0,0 +1,144 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+#pragma once
+
+#include "universetypes.h"
+#include "types.h"
+
+#include <entt/core/family.hpp>
+#include <entt/core/any.hpp>
+
+#include <Corrade/Containers/StridedArrayView.h>
+
+#include <optional>
+#include <variant>
+#include <vector>
+
+namespace osp::universe
+{
+
+using ccomp_family_t = entt::family<struct ccomp_type>;
+using ccomp_id_t = ccomp_family_t::family_type;
+
+// Coordinate component types (CComp). ccomp_family_t generates IDs for these
+// at runtime, used as indices for CoordinateSpace::m_components
+struct CCompX { using datatype_t = spaceint_t; };
+struct CCompY { using datatype_t = spaceint_t; };
+struct CCompZ { using datatype_t = spaceint_t; };
+struct CCompSat { using datatype_t = Satellite; };
+
+template <typename CCOMP_T>
+using ViewCComp_t = Corrade::Containers::StridedArrayView1D<typename CCOMP_T::datatype_t>;
+
+template<typename CCOMP_T>
+constexpr ccomp_id_t ccomp_id() noexcept
+{
+    return ccomp_family_t::type<CCOMP_T>;
+}
+
+using CoordinateView_t = std::optional<Corrade::Containers::StridedArrayView1D<void>>;
+
+/**
+ * @brief Stores positions, velocities, and other related data for Satellites,
+ *        and exposes them through read-only strided array views.
+ *
+ * CoordinateSpaces must be managed by some external system. They can store any
+ * kind of coordinate data, such as Cartesian XYZ or orbital parameters.
+ *
+ * As part of ECS, this is a way to store common components in separate buffers.
+ * that can each be managed by a specific system (simultaneously too).
+ * aka: a non-EnTT but still ECS component pool
+ *
+ */
+struct CoordinateSpace
+{
+    using SatToAdd_t = std::tuple<Satellite, Vector3g, Vector3>;
+
+    enum class ECmdOp : uint8_t { Add, Set };
+    enum class ECmdVar : uint8_t { Position, Velocity };
+    using CmdValue_t = std::variant<Vector3, Vector3g>;
+
+    using Command_t = std::tuple<ECmdOp, ECmdVar, CmdValue_t>;
+
+    /**
+     * @brief Request to add a Satellite to this coordinate space
+     *
+     * Every coordinate space must be able to accept Satellites by a position
+     * and a velocity as a common interface. If the CoordinateSpace uses a
+     * non-Cartesian coordinate system like Kepler orbits, then it must be
+     * converted.
+     *
+     * @param sat [in] Satellite to add
+     * @param pos [in] Position in this coordinate system
+     * @param vel [in] Velocity of satellite
+     */
+    void add(Satellite sat, Vector3g pos, Vector3 vel)
+    {
+        m_toAdd.emplace_back(sat, pos, vel);
+    }
+
+    /**
+     * @brief TODO - commands to move and accelerate satellites
+     *
+     * @param cmd
+     */
+    void command(Command_t&& cmd)
+    {
+        m_commands.emplace_back(std::move(cmd));
```

Comment:
For situations where there's no meaningful difference between the implementations of by const-ref and by-rvalue-ref, i normally just take the variable in question by value.

It has the advantage of, depending on the specific ABI of the platform, providing the compiler more opportunities for passing the data in registers.